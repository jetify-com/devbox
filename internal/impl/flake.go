package impl

import (
	"io"
	"os"
	"os/exec"
	"path/filepath"

	"github.com/pkg/errors"
)

func (d *Devbox) installNixProfileFlakeCommand(profileDir string) *exec.Cmd {
	cmd := exec.Command(
		"nix", "profile", "install",
		"--profile", profileDir,
		"--extra-experimental-features", "nix-command flakes",
		filepath.Join(d.configDir, ".devbox/gen/flake/"), // installables
	)
	if d.hasDevboxLockFile() {
		_ = d.copyDevboxLockToFlakeLock()
		cmd.Args = append(cmd.Args, "--no-write-lock-file")
	} else {
		cmd.Args = append(cmd.Args, "--recreate-lock-file")
	}
	return cmd
}

func (d *Devbox) copyFlakeLockToDevboxLock() error {

	flakeLock, err := os.Open(d.flakeLockPath())
	if err != nil {
		return errors.WithStack(err)
	}
	devboxLock, err := os.Create(d.devboxLockPath())
	if err != nil {
		return errors.WithStack(err)
	}
	_, err = io.Copy(devboxLock, flakeLock)
	if err != nil {
		return errors.WithStack(err)
	}
	return nil
}

func (d *Devbox) copyDevboxLockToFlakeLock() error {
	if !d.hasDevboxLockFile() {
		return errors.New("devbox.lock file does not exist")
	}

	devboxLock, err := os.Open(d.devboxLockPath())
	if err != nil {
		return errors.WithStack(err)
	}
	flakeLock, err := os.Create(d.flakeLockPath())
	if err != nil {
		return errors.WithStack(err)
	}
	_, err = io.Copy(flakeLock, devboxLock)
	if err != nil {
		return errors.WithStack(err)
	}
	return nil
}

func (d *Devbox) hasDevboxLockFile() bool {
	if _, err := os.Stat(d.devboxLockPath()); err != nil {
		return false
	}
	return true
}

func (d *Devbox) devboxLockPath() string {
	return filepath.Join(d.configDir, "devbox.lock")
}

func (d *Devbox) flakeLockPath() string {
	return filepath.Join(d.configDir, ".devbox/gen/flake/flake.lock")
}
