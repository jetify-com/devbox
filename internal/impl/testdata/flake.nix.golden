{
  description = "A devbox shell";

  inputs = {
    nixpkgs.url = "https://github.com/nixos/nixpkgs/archive/b9c00c1d41ccd6385da243415299b39aa73357be.tar.gz";
    flake-utils.url = "github:numtide/flake-utils";
  };

  outputs = { 
    self,
    nixpkgs,
    flake-utils 
  }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = (import nixpkgs {
          inherit system;
          config.allowUnfree = true;
        });
        php = pkgs.php.withExtensions ({ enabled, all }: enabled ++ (with all; [ blackfire ]));
        php81Packages.composer = php.packages.composer;
      in
      {
        devShell = pkgs.mkShell {
          buildInputs = with pkgs; [
            php
            php81Packages.composer
            php
            php81Packages.composer
            php81Extensions.blackfire
            flyctl
            postgresql
            tree
            git
            zsh
            openssh
            vim
            sqlite
            jq
            delve
            ripgrep
            shellcheck
            terraform
            xz
            zstd
            gnupg
            go_1_20
            python3
            graphviz
          ];
        };
      }
    );
}
